File: src/common/Character.ts
import { faker } from '@faker-js/faker';
import { Stats } from './Abilities';
import { Items } from './Items';

export interface ICharacter {
  name: string;
  level: number;
}

export default abstract class Character implements ICharacter, Stats {
  name!: string;
  level: number = 1;
  health!: number;
  speed!: number;
  damage!: number;
  defense!: number;
  magia!: number;
  move!: number;
  mana!: number;
  items: Items[] = [];

  constructor(init: Omit<ICharacter, 'name' | 'level'>) {
    Object.assign(this, init);
    this.name = faker.internet.userName();
  }

  public abstract attack(): number;
  public abstract levelUp(): void;
  public abstract activateSkill(): void;
}

File: abstracts/character.py

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import List, Optional
from interfaces.Item import Item

from interfaces.stats import Stats


class CharacterICharacter:
    name: str
    level: int


@dataclass
class Character(ABC, Stats):
    name: str
    level: int = 1
    items: Optional[List[Item]] = None

    @abstractmethod
    def attack() -> int:
        raise NotImplementedError("attack is not implemented")

    @abstractmethod
    def level_up() -> None:
        raise NotImplementedError("level_up is not implemented")

    @abstractmethod
    def activate_skill() -> None:
        raise NotImplementedError("activate_skill is not implemented")
